### PR TITLE
Add tests from messageformat@4.0.0-7

### PR DESCRIPTION
**Pull request by eemeli**
Update the test suite with new and changed tests to match the latest spec, as a copy from https://github.com/messageformat/messageformat/tree/messageformat@4.0.0-7/packages/mf2-messageformat/src/__fixtures.

These add tests for markup-close with options, `:date`, `:time`, and `:datetime`, as well as various corner cases encountered during development.

### CODE REVIEW

File: test/test-functions.json
Comment (aphillips) [inline test/test-functions.json:7]: I believe the spec says this should be an invalid expression error. Your code might call this a `RangeError`. Fix?

Comment (eemeli) [reply test/test-functions.json:7]: Ah, true. I currently don't add a wrapper for the built in JS RangeError that's the cause here, as the `:number` spec requires.

File: test/syntax-errors.json
Comment (gibson042) [inline test/syntax-errors.json:22]: These don't seem like "missing space".
```suggestion
  "missing equals {:func @bar@baz}",
  "missing missing quotes {:func @bar=42@baz}",
  "reserved {+reserved@bar}",
  "private {&private@bar}",
```

**Comment by mihnita** [comment]
Suggestion (does not have to be in this PR): unify schemas?

A good target would be the one used for functions.

For example `syntax-errors`:
```
{
    "errors" : [
        {
           "src": "missing space {|foo|:func}",
        }
    ]
}
```

And `core`:
```
{
    "errors" : [
       ... the current cases
   ]
}
```

File: test/syntax-errors.json
Comment (eemeli) [reply test/syntax-errors.json:22]: All of those can be turned into valid syntax with the addition of a single space:
```suggestion
  "missing space {:func @bar @baz}",
  "missing space {:func @bar=42 @baz}",
  "missing space {+reserved @bar}",
  "missing space {&private @bar}",
```
(the above is not intended as an actual suggestion, just to show a nice diff)

**Comment by eemeli** [comment]
> Suggestion (does not have to be in this PR): unify schemas?
> 
> A good target would be the one used for functions.

Not really sure what you mean here. At least for now, all of the `syntax-errors.json` entries are spec'd just to produce a syntax or data model error, and formatting those messages should never even get to calling any functions. The errors in the other files are runtime errors.

**Comment by mihnita** [comment]
> Not really sure what you mean here. 
> At least for now, all of the syntax-errors.json entries are spec'd just to produce a syntax or data model error

What I mean is that `syntax-errors.json` contains a list of strings that should fail parsing.
But there is no way to tell what is the error they fail with.

For example:
```json
[
  ...
  "bad {:placeholder option=x option=x}",
  ...
  ".input {$foo} .input {$foo} {{_}}"
  ...
]
```
vs.
```json
{
  "syntax-errors": [
    ...
  ],
  "data-model-errors": [      
    {
      "src": "bad {:placeholder option=x option=x}",
      "errors": [{ "type": "duplicate-option-name" }]
    }
    {
      "src": ".input {$foo} .input {$foo} {{_}}",
      "errors": [{ "type": "duplicate-declaration" }]
    }     
  ]
}
```
---

The second benefit is that there is one single schema for all test files.
Making it easier to parse (at least in some languages :-).

**Comment by eemeli** [comment]
@aphillips Could this be merged, so that further work on the tests can build on top of it, such as #707 and #708? I understood that to be the conclusion reached during the last call.

**Review by mihnita — APPROVED** [review]

File: test/syntax-errors.json
Comment (aphillips) [reply test/syntax-errors.json:22]: Technically, though, they are invalid name tests. You created them by removing non-optional spaces, of course.

**Comment by aphillips** [comment]
Per our conversation, we were waiting on people not named @aphillips and @eemeli to review before merging. But we're holding up additional work by retaining this as a PR.

**Pull request merged**
